### PR TITLE
Implement function to compute the canonical word H(s) of an A/B configuration s

### PR DESCRIPTION
**Pull request by morxa**
This implements a canonical word representation H(s) of an A/B configuration s, closely following Bouyer et al., 2006, p. 460.

The TAStates of s are first expanded into triples (location, clock, valuation) (one for each clock), and then merged with the pairs from the ATAConfiguration. The resulting set is then partitioned according to the fractional part of the clock valuations. Then, each tuple is regionalized by replacing the clock valuation with the respective region index.

### CODE REVIEW

File: src/synchronous_product/include/synchronous_product/synchronous_product.h [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+{
+	std::set<ABSymbol<Location, ActionType>> g;
+	// Insert ATA configurations into g.
+	std::copy(ata_configuration.begin(), ata_configuration.end(), std::inserter(g, g.end()));
+	// Insert TA configurations into g.
+	for (const auto &[clock_name, clock_value] : ta_configuration.second) {
+		g.insert(std::make_tuple(ta_configuration.first, clock_name, clock_value));
+	}
+	// Sort into partitions by the fractional parts.
+	std::map<double, std::set<ABSymbol<Location, ActionType>>> partitioned_g;
```
Comment (modass) [inline src/synchronous_product/include/synchronous_product/synchronous_product.h:106]: I think double should be ClockValuation, right?

File: src/synchronous_product/include/synchronous_product/synchronous_product.h [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+		  [&](const ABSymbol<Location, ActionType> &w) -> ABRegionSymbol<Location, ActionType> {
+			  if (std::holds_alternative<TAState<Location>>(w)) {
+				  const TAState<Location> &s = std::get<TAState<Location>>(w);
+				  return TARegionState<Location>(std::get<0>(s),
+				                                 std::get<1>(s),
+				                                 regionSet.getRegionIndex(std::get<2>(s)));
+			  } else {
+				  const ATAState<ActionType> &s = std::get<ATAState<ActionType>>(w);
+				  return ATARegionState<ActionType>(s.first, regionSet.getRegionIndex(s.second));
+			  }
```
Comment (modass) [inline src/synchronous_product/include/synchronous_product/synchronous_product.h:129]: The inner if-switch can be replaced by a visitor - do we want this? I will make one suggestion and you can check, whether you like it - if so, I can add further.

Comment (modass) [reply src/synchronous_product/include/synchronous_product/synchronous_product.h:129]: Never mind, the visitor is overly verbose.

**Review by modass — APPROVED** [review]
Works out fine for me

File: src/synchronous_product/include/synchronous_product/synchronous_product.h (hunk 1 above)
Comment (morxa) [reply src/synchronous_product/include/synchronous_product/synchronous_product.h]: Yes, thanks for the fix!

**Pull request merged**